patches/server/0965-Make-banners-modifiable-with-both-BannerMeta-and-Blo.patch

```diff
@@ -0,0 +1,270 @@
+         super.serialize(builder);
+ 
++        if (true) return builder; // Paper - Let CraftMetaBlockState serialize for us
+         if (this.base != null) {
+             builder.put(BASE.BUKKIT, this.base.toString());
+         }
+@@ -206,6 +277,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+     int applyHash() {
+         final int original;
+         int hash = original = super.applyHash();
++        if (true) return original; // Paper - CraftMetaBlockState's hash is good enough
+         if (this.base != null) {
+             hash = 31 * hash + this.base.hashCode();
+         }
+@@ -217,6 +289,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+ 
+     @Override
+     public boolean equalsCommon(CraftMetaItem meta) {
++        if (true) return super.equalsCommon(meta); // Paper - Only check CraftMetaBlockState's data
+         if (!super.equalsCommon(meta)) {
+             return false;
+         }
+@@ -230,11 +303,13 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+ 
+     @Override
+     boolean notUncommon(CraftMetaItem meta) {
++        if (true) return super.notUncommon(meta) && (meta instanceof CraftMetaBanner || this.blockEntityTag == null);  // Paper - Only check CraftMetaBlockState's data
+         return super.notUncommon(meta) && (meta instanceof CraftMetaBanner || (this.patterns.isEmpty() && this.base == null));
+     }
+ 
+     @Override
+     boolean isEmpty() {
++        if (true) return super.isEmpty(); // Paper - Only check CraftMetaBlockState's data
+         return super.isEmpty() && this.patterns.isEmpty() && this.base == null;
+     }
+ 
+@@ -246,7 +321,7 @@ public class CraftMetaBanner extends CraftMetaItem implements BannerMeta {
+     @Override
+     public CraftMetaBanner clone() {
+         CraftMetaBanner meta = (CraftMetaBanner) super.clone();
+-        meta.patterns = new ArrayList<>(this.patterns);
++        // meta.patterns = new ArrayList<>(this.patterns); // Paper - Only copy CraftMetaBlockState's data
+         return meta;
+     }
+ }
+diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
+index 1957e82b67b4aa301d12468d5a81d96e621dab49..8413721ba51191555d66dedb14cbb2b6ec99e61e 100644
+--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
++++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaBlockState.java
+@@ -40,7 +40,7 @@ public class CraftMetaBlockState extends CraftMetaItem implements BlockStateMeta
+             Material.BLACK_SHULKER_BOX
+     );
+ 
+-    private static final Set<Material> BLOCK_STATE_MATERIALS = Sets.newHashSet(
++    static final Set<Material> BLOCK_STATE_MATERIALS = Sets.newHashSet( // Paper - Change it to package-private, so we can use that in CraftMetaBanner
```

Comment:
Don't do this, do the reverse. Make the list of banners in CraftMetaBanner not private and add it to the list here, same way the list of shulker boxes is.

Reply:
I actually tried to do this first, but since BannerBeta extends CraftBlockStateMeta, the static initializer of CraftBlockStateMeta tries to add the uninitialized Set of BannerMeta (which is null at that moment), making that entire thing screw up